Cesium.CodeGen/ConstantEvaluator.cs

```diff
@@ -2,71 +2,84 @@
 using Cesium.CodeGen.Ir.Expressions.BinaryOperators;
 using Cesium.CodeGen.Ir.Expressions.Constants;
 using Cesium.Core;
+using System.Diagnostics;
 
 namespace Cesium.CodeGen;
 
 internal static class ConstantEvaluator
 {
     public static IConstant GetConstantValue(IExpression expression)
+    {
+        var result = TryGetConstantValue(expression);
+        if (result.ErrorMessage is not null)
+        {
+            throw new CompilationException(result.ErrorMessage);
+        }
+
+        Debug.Assert(result.Constant != null);
+        return result.Constant;
+    }
+
+    public static (string? ErrorMessage, IConstant? Constant) TryGetConstantValue(IExpression expression)
     {
         switch (expression)
         {
             case ConstantLiteralExpression literal:
-                return literal.Constant;
+                return (null, literal.Constant);
 
             case UnaryOperatorExpression unOp:
-            {
-                var constant = GetConstantValue(unOp.Target);
+                {
+                    var constant = GetConstantValue(unOp.Target);
 
-                if (constant is not IntegerConstant constInt)
-                    throw new CompilationException("Evaluated constant is not an integer");
+                    if (constant is not IntegerConstant constInt)
+                        return ("Evaluated constant is not an integer", null);
 
                     return unOp.Operator switch
                     {
-                        UnaryOperator.Negation => new IntegerConstant(-constInt.Value),
-                        UnaryOperator.BitwiseNot => new IntegerConstant(~constInt.Value),
-                        UnaryOperator.LogicalNot => new IntegerConstant(constInt.Value != 0 ? 0 : 1),
-                        UnaryOperator.AddressOf or UnaryOperator.Indirection => throw new CompilationException($"Operator {unOp.Operator} is not compile-time evaluable"),
-                        _ => throw new ArgumentOutOfRangeException($"Invalid unary operator {unOp.Operator}"),
+                        UnaryOperator.Negation => (null, new IntegerConstant(-constInt.Value)),
+                        UnaryOperator.BitwiseNot => (null, new IntegerConstant(~constInt.Value)),
+                        UnaryOperator.LogicalNot => (null, new IntegerConstant(constInt.Value != 0 ? 0 : 1)),
+                        UnaryOperator.AddressOf or UnaryOperator.Indirection => ($"Operator {unOp.Operator} is not compile-time evaluable", null),
+                        _ => ($"Invalid unary operator {unOp.Operator}", null),
```

Comment:
I suggest keeping this one as an `ArgumentOutOfRangeException`.

While the others are, indeed, compilation errors — something that the user should be able to correct, this one in compiler implementation issue, and should not be mixed with the others. It should be thrown immediately, on sight.